src/Approximations/decompositions.jl

```diff
@@ -1,60 +1,37 @@
-"""
-    default_block_structure(S::LazySet, set_type::Type{<:LazySet})::AbstractVector{Int}
-
-Compute the default block structure.
-
-### Input
-
-- `S`        -- set
-- `set_type` -- target set type
-
-### Output
-
-A vector representing the block structure, such that:
-
-- If the target `set_type` is an interval, the default is blocks of size 1.
-- Otherwise, the default is blocks of size 2. Depending on the dimension,
-  the last block has size 1 or 2.
-"""
-@inline function default_block_structure(S::LazySet, set_type::Type{<:LazySet})::AbstractVector{Int}
-    n = dim(S)
-    if set_type == Interval
-        return fill(1, n)
-    else
-        if n % 2 == 0
-            return fill(2, div(n, 2))
-        else
-            res = fill(2, div(n+1, 2))
-            res[end] = 1
-            return res
-        end
+# compute a uniform block partition
+@inline function uniform_partition(n::Int, block_size::Int)
+    m = div(n, block_size)
+    r = n % block_size
+    res = Vector{UnitRange{Int}}(undef, r > 0 ? m + 1 : m)
+    k = 1
+    @inbounds for i in 1:m
+        l = k + block_size - 1
+        res[i] = k:l
+        k = l + 1
     end
+    if r > 0
+        res[m+1] = k:n
+    end
+    return res
 end
 
 """
-    decompose(S::LazySet{N};
-              [set_type]::Type{<:Union{HPolygon, Hyperrectangle, Interval}}=Hyperrectangle,
-              [ε]::Real=Inf,
-              [blocks]::AbstractVector{Int}=default_block_structure(S, set_type),
-              [block_types]::Dict{Type{<:LazySet}, AbstractVector{<:AbstractVector{Int}}}(),
-              [directions]::Union{Type{<:AbstractDirections}, Nothing}=nothing
-             )::CartesianProductArray where {N<:Real}
+    decompose(S::LazySet{N},
+              partition::AbstractVector{<:AbstractVector{Int}},
+              block2oa
+             )::CartesianProductArray{N} where {N<:Real}
 
 Decompose a high-dimensional set into a Cartesian product of overapproximations
 of the projections over the specified subspaces.
 
 ### Input
 
-- `S`           -- set
-- `set_type`    -- (optional, default: `Hyperrectangle`) type of set approximation
-                   for each subspace
-- `ε`           -- (optional, default: `Inf`) error bound for polytopic approximation
-- `blocks`      -- (optional, default: [2, …, 2] or [1, …, 1] if `set_type` is an interval)
-                   block structure - a vector with the size of each block
-- `block_types` -- (optional, default: Interval for 1D and Hyperrectangle
-                   for mD blocks) a mapping from set types to blocks
-- `directions`  -- (optional, default: `nothing`) template direction type, or
-                   `nothing`
+- `S`         -- set
+- `partition` -- vector of blocks (i.e., of vectors of integers) (see the Notes
+                 below)
+- `block2oa`  -- mapping from block indices in `partition` to a corresponding
```

Comment:
can we use a more descriptive name for `block2oa`? like `options`, `block_options`, `set_types`, `block_types`,..